### PR TITLE
Country::Factory

### PR DESCRIPTION
**Pull request by IhateTrains**
#70

### CODE REVIEW

File: ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		country->currencies.gold = currenciesFromBloc.getGold();
+		country->currencies.stability = currenciesFromBloc.getStability();
+		country->currencies.tyranny = currenciesFromBloc.getTyranny();
+		country->currencies.war_exhaustion = currenciesFromBloc.getWarExhaustion();
+		country->currencies.aggressive_expansion = currenciesFromBloc.getAggressiveExpansion();
+		country->currencies.political_influence = currenciesFromBloc.getPoliticalInfluence();
+		country->currencies.military_experience = currenciesFromBloc.getMilitaryExperience();
+	});
+	registerKeyword("capital", [this](std::istream& theStream) {
+		auto capitalLongLong = commonItems::getULlong(theStream);
```
Comment (Idhrendur) [inline ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp:55]: Just 'capital' is fine

File: ImperatorToCK3/Source/Imperator/Countries/Country.h [hunk 2]
```diff
@@ -55,9 +58,7 @@ namespace Imperator
 			void setCK3Title(const std::shared_ptr<CK3::Title>& theTitle) { ck3Title = theTitle; }
 
 		private:
-			void registerKeys();
-
-			unsigned long long countryID = 0;
+			unsigned long long ID = 0;
```
Comment (Idhrendur) [inline ImperatorToCK3/Source/Imperator/Countries/Country.h:61]: I have a mild preference for uint64_t as it doesn't require looking up or memorizing the size of a long long. But it's not a big deal either way. Requires including cstdint

File: ImperatorToCK3Tests/ImperatorWorldTests/Countries/CountryTests.cpp [hunk 3]
```diff
@@ -233,29 +234,29 @@ TEST(ImperatorWorld_CountryTests, color3DefaultsToNullopt)
...
 	ASSERT_FALSE(theCountry.getColor3());
 }
 
 TEST(ImperatorWorld_CountryTests, correctCountryRankIsReturned)
 {
 	std::stringstream input;
-	const Imperator::Country theCountry1(input, 1);
-	
-	Imperator::Country theCountry2(input, 2);
+	auto theCountry1 = *Imperator::Country::Factory().getCountry(input, 1);
```
Comment (Idhrendur) [inline ImperatorToCK3Tests/ImperatorWorldTests/Countries/CountryTests.cpp:245]: If you're creating multiple countries in a test, it's worth considering storing the factory as a variable and using in each case.

**Review by Idhrendur — APPROVED** [review]

File: ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+#include "ParserHelpers.h"
+#include "CommonRegexes.h"
+#include "CountryFactory.h"
+#include "CountryCurrencies.h"
+#include "CountryName.h"
+
+
+Imperator::Country::Factory::Factory()
+{
+	registerSetter("tag", country->tag);
```
Comment (Idhrendur) [inline ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp:11]: Setters don't work with factories, unfortunately. The captured variable gets set at the time the constructor runs, when the country is an invalid pointer, and it doesn't get updated when you create the country during the getCountry() call. This caused a segfault. I don't know why that's different than having the lambda be there directly, but it is.

**Review by Idhrendur — CHANGES_REQUESTED** [review]

File: ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp (hunk 4 above)
Comment (IhateTrains) [reply ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp:11]: That's unfortunate. Your explanation makes sense. The difference is probably that in the registerSetter the lamdba captures the variable, while in registerKeyword the lambda usually captures `this`.

File: ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp (hunk 1 above)
Comment (IhateTrains) [reply ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp]: Renamed to `capitalProvID` for less confusion with the member.

File: ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp (hunk 4 above)
Comment (IhateTrains) [reply ImperatorToCK3/Source/Imperator/Countries/CountryFactory.cpp]: Changed to registerKeyword ✔️

**Review by Idhrendur — APPROVED** [review]

**Pull request merged**